hy/compiler.py

```diff
@@ -121,6 +122,13 @@ def _dec(fn):
     return _dec
 
 
+def builds_if(_type, condition):
+    if condition:
+        return builds(_type)
+    else:
+        return lambda fn: fn
```

Comment:
No, it cannot (although I'm making a mental note to ask myself that good question next time I find myself writing the identity function).

The thought process here was that `builds_if` is going to be a decorator that takes arguments, which one can think of as a function that returns the "actual" decorator that will wrap the function being decorated by the decorator-with-arguments. The indended behavior is that if `condition` is truthy, we want a function decorated with `@builds_if("★", condition)` (this example imagines ourselves defining a [\u2605 Black Star](http://www.fileformat.info/info/unicode/char/2605/index.htm) operator, perhaps for some hypothetical far-future version of Python) to behave as if it were decorated with `@builds("★")`, so in that case we return the decorator that we want: `builds("★")`. But if `condition` is falsey, we want the `@builds_if("★", condition)` decorator to be a no-op (that just passes the function being decorated on to the next decorator in the decorator-chain as if it were never there). But the no-op decorator is the identity function; as the pasted output below shows, we don't want to return `_type` because `_type` is a string in the actual usage of `@builds` and `@builds_if`&mdash;

```
(hy) zmd@ExpectedReturn:~/Code/hy/hy$ git diff
diff --git a/hy/compiler.py b/hy/compiler.py
index 8eb2291..7209f82 100644
--- a/hy/compiler.py
+++ b/hy/compiler.py
@@ -123,10 +123,11 @@ def builds(_type):
 
 
 def builds_if(_type, condition):
+    print("MY DEBUG MARKER A", _type, condition)
     if condition:
         return builds(_type)
     else:
-        return lambda fn: fn
+        return _type
 
 
 class Result(object):
(hy) zmd@ExpectedReturn:~/Code/hy/hy$ hy
MY DEBUG MARKER A @ False
Traceback (most recent call last):
  File "/home/zmd/Code/hy/bin/hy", line 9, in <module>
    load_entry_point('hy==0.10.1', 'console_scripts', 'hy')()
  File "/home/zmd/Code/hy/lib/python3.4/site-packages/setuptools-0.9.6-py3.4.egg/pkg_resources.py", line 378, in load_entry_point
  File "/home/zmd/Code/hy/lib/python3.4/site-packages/setuptools-0.9.6-py3.4.egg/pkg_resources.py", line 2566, in load_entry_point
  File "/home/zmd/Code/hy/lib/python3.4/site-packages/setuptools-0.9.6-py3.4.egg/pkg_resources.py", line 2260, in load
  File "/home/zmd/Code/hy/hy/hy/__init__.py", line 37, in <module>
    import hy.importer  # NOQA
  File "/home/zmd/Code/hy/hy/hy/importer.py", line 22, in <module>
    from hy.compiler import hy_compile, HyTypeError
  File "/home/zmd/Code/hy/hy/hy/compiler.py", line 370, in <module>
    class HyASTCompiler(object):
  File "/home/zmd/Code/hy/hy/hy/compiler.py", line 1651, in HyASTCompiler
    @checkargs(min=2)
TypeError: 'str' object is not callable
```